documentation/Architecture/The-Data-Processing-Pipeline.md

```diff
@@ -1,43 +1,131 @@
+> 2) Having both provides developers with more flexibility on how they structure their plugins
 
-The Source Parser is bound to a procesor which is created by a CDS in order to:
-* read in raw data from the Data Source 
-* create a standardized data format
+From an interface perspective, a `SourceParser` has the following properties:
+* A `SourceParser` has a `string` identifier
+* A `SourceParser` is bound to a specific `CustomDataProcesor`
+* A `SourceParser` must, at some point, process the data sources originally given to the `CustomDataSource` which  
+  created the `CustomDataProcessor` the parser is bound to 
+* A `SourceParser` will, *while processing its data sources*, emit each individual data point contained 
+  within the data sources it is processing
 
-The standardized data is then sent to a data cooker.
+The data points a `SourceParser` emits are called __events__, and each __event__ must have a *key*. Conceptually, 
+the emitted events are `(key, value)` pairs contained within a single class, where each value is a data point and its 
+key is some property of the data point. 
 
-# Data Cooker
+A `SourceParser` emits each event one at a time. __It is highly encouraged to make sure a source parser 
+only parses/iterates the data inside its data sources once, emitting each event as soon as it is 
+encountered.__
 
-The Data Cooker consumes events sent from the Source Parser and specifies the data keys. 
-This allows for the standardized data to be read into a C# data structure and allows programmatic access to any data.
+## Data Cookers
 
-The platform Data Cookers and Tables are built on is extensible by default. This means Data Cookers can not only be referenced by 
-eachother but also referenced in other binaries and other projects.
+A `DataCooker` is a class whose purpose is to implement the data transformations outlined at the beginning of this document. 
+From an interface perspective, a `DataCooker` has the following properties:
+* A `DataCooker` has a `DataCookerPath` that can 
+    1) be recreated by other modules (or external assemblies) and 
+    2) be used to reference that `DataCooker` (even without access to that `DataCooker`'s source code)
+* A `DataCooker` consumes data from from either a *single source parser* or *one or more data cookers*
+* A `DataCooker` processes (cooks) this data in some way
+* A `DataCooker` exposes zero or more __data output__ properties
+
+`DataOutput`s are the key component of data cookers, and are what achieve the last two goals outlined at the beginning of this document. 
+The SDK is aware of every `DataOutput` a given `DataCooker` advertises, and any assembly with access to an instance of 
+the SDK with a `DataCooker` loaded can query that cooker for any of its `DataOutput`s. For instance, a `Table` can 
+query transformed data outputted by a `DataCooker` defined inside its own assembly:
+
+<img src=".attachments/cooker_query.svg" width="800">
```

Comment:
Consider an arrow bouncing off the SDK for a visual to either replace or complement the "(through the SDK)"

Reply:
I was trying to think of ways to visually represent it going through the SDK cleanly but didn't find any that looked nice. A bouncing arrow is a good idea! 👍